sycl/include/sycl/types.hpp

```diff
@@ -2165,18 +2174,37 @@ template <typename T, int N, typename V> struct VecStorage {
   static_assert(!std::is_same_v<V, void>, "Incorrect data type for sycl::vec");
 };
 
-#ifdef __SYCL_USE_EXT_VECTOR_TYPE__
+#ifdef __SYCL_DEVICE_ONLY__
+// device always has ext vector support, but for huge vectors
+// we switch to std::array, so that we can use a smaller alignment (64)
+// this is to support MSVC, which has a max of 64 for direct params.
 template <typename T, int N> struct VecStorageImpl {
-  using DataType = T __attribute__((ext_vector_type(N)));
+  static constexpr size_t Num = (N == 3) ? 4 : N;
+  static constexpr size_t Sz = Num * sizeof(T);
+  using DataType =
+      typename std::conditional<Sz <= 64, T __attribute__((ext_vector_type(N))),
+                                std::array<T, Num>>::type;
+  using VectorDataType = T __attribute__((ext_vector_type(N)));
 };
-#else
+#else // __SYCL_DEVICE_ONLY__
 // When ext_vector_type is not available, we rely on cl_* types from CL/cl.h
 // to represent vec storage.
+#ifdef __HAS_EXT_VECTOR_TYPE__
```

Comment:
I believe we should always use `std::array` on the host, otherwise we may break when translation units compiled using different compilers.

Reply:
Ok, I've made the requested change to std::array only on host  and I removed the now irrelevant `__NO_EXT_VECTOR_TYPE_ON_HOST__` macro.